### PR TITLE
Autofill previously entered accounting and routing numbers

### PR DESCRIPTION
**Pull request by Jag96**
@roryabraham will you please review this?

### Details
This PR updates the Add Bank Account flow to save the accountNumber and routingNumber if the user exits the flow after setting those values.

### Fixed Issues

$ https://github.com/Expensify/App/issues/4476

### Tests/QA
1. Log into an account w/ a workspace and go to Settings->Workspace->Expensify Card and hit `Get Started`
2. Fill out values for the accounting number and routing number and hit `Save & Continue`
3. Close the modal then hit `Get Started` again, confirm routing number and account number are saved
4. Change the routing and account numbers, then hit `Save & Continue`
5. Close the modal then hit `Get Started` again. Verify that the routing and account number have been updated.
6. Hit `Save & Continue` again.
7. Fill out the company information step and hit save and continue
8. Close the modal and hit `Get Started` again, confirm the values are pre-populated
9. Sign out of the account and sign back in, confirm that the values are pre-populated (this will only work if you save on the Company Information step)

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Web/Desktop
| Starting Flow | Already saved at Company Info |
|---|---|
|  |  |

#### Mobile
| Starting Flow | Already saved at Company Info |
|---|---|
|  |  |' width='400'/> |

### CODE REVIEW

**Review by Julesssss — DISMISSED** [review]

**Review by Julesssss — COMMENTED** [review]
Just noticed a case that I think might be a failure. It's a bit of an edge case, but we will definitely have customers who might not notice that the old numbers overwrote the newer details

- Run test steps 1 & 2
- Now, reopen the 'Add Bank Account' step but this time change the acc numbers
- Hit save & continue, then close the modal
- Reopen the 'Add Bank Account' step
- Notice that the OLD numbers are used -- this should instead show the last entered numbers (666666666 instead of 555555555 in the screencast)


https://user-images.githubusercontent.com/10736861/128697070-2bbd3e53-fce4-493f-8507-c3acfefbefea.mov

**Comment by Jag96** [comment]
Looking into this, it seems like there is an Onyx.merge call that isn't working properly in this flow. In [this code](https://github.com/Expensify/App/blob/9adaa96fc435de40f43773435931a960ab44175b/src/libs/actions/BankAccounts.js#L622) the `newACHData` is being passed with a routingNumber `123123123`, but after the merge resolves the value in Onyx is still set to the old value `555555555`. Going to have a look and see why this is happening.

image

**Comment by Jag96** [comment]
Ok, looks like the fix here is to just wait until the first merge call finishes before letting the 2nd one happen since the 2nd one is overwriting the first with the previous values. Testing and updating.

**Comment by Jag96** [comment]
@roryabraham @Julesssss updated to fix the edge case there, good find!

**Comment by roryabraham** [comment]
Updated the testing steps to include the edge case you found @Julesssss

File: src/libs/actions/BankAccounts.js
Comment (roryabraham) [inline src/libs/actions/BankAccounts.js:621]: This is weird, I think it should maybe be like this instead?

```javascript
// First, optimistically update the reimbursement account in Onyx
Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {
    loading: false,
    achData: {...newACHData},
})

    // Then update it in the API
    .then(() => API.BankAccount_SetupWithdrawal(newACHData))
    .then((response) => {
        // Do other stuff
    });
```

**Review by roryabraham — CHANGES_REQUESTED** [review]

File: src/libs/actions/BankAccounts.js
Comment (roryabraham) [reply src/libs/actions/BankAccounts.js:621]: Also probably want to add a `catch` block to undo the optimistic change if the API call doesn't work for some reason.

Comment (roryabraham) [reply src/libs/actions/BankAccounts.js:621]: Or also undo the optimistic change if the API returns an error response.

Comment (Jag96) [reply src/libs/actions/BankAccounts.js:621]: I agree that's the ideal flow, but I'm a bit hesitant to make that change right now since it will affect every step in the VBA flow. We currently have https://github.com/Expensify/Expensify/issues/173363 to ensure we go back and refactor the flow to make it easier to understand and less buggy, perhaps we can make this improvement there?

Comment (roryabraham) [reply src/libs/actions/BankAccounts.js:621]: > Also probably want to add a catch block to undo the optimistic change if the API call doesn't work for some reason.

This is actually already done [here](https://github.com/Expensify/App/blob/f5bb5345f0319deeec2d9d64d78363a3daecadcf/src/libs/actions/BankAccounts.js#L756), so we should be good there.

> Or also undo the optimistic change if the API returns an error response.

We have some minimal error handling set up, but adjusting this can be out-of-scope for this PR. Even in its current form, this PR does does this before checking the API response status code:

```javascript
Onyx.merge(ONYXKEYS.REIMBURSEMENT_ACCOUNT, {
    loading: false,
    achData: {...newACHData},
})
```

So doing that before the API call or after should make no difference to the error handling.

All this is to say that the only change that needs to be made is the one from my first comment in this thread, which imo is pretty simple and safe to make. The only thing happening between the time when I'm suggesting we do the `Onyx.set` and when you're currently doing the `Onyx.set` is an API call, and the API call has no dependence on the Onyx data, so we should be good. We just have our promise chain out-of-order in this PR in its current form.

Comment (Jag96) [reply src/libs/actions/BankAccounts.js:621]: > We have some minimal error handling set up, but adjusting this can be out-of-scope for this PR.

Sounds good, I'll have a look at updating just the API call order and doing some extra testing to make sure everything still works.

Comment (Jag96) [reply src/libs/actions/BankAccounts.js:621]: Did some testing here and making that update will result in the UI being blocked on the API call since the UI updates all happen in the `.then` handler after the API call is done, so the loading spinner isn't shown and the screen looks frozen. To update this, we'd have to do everything optimistically outside of the then function, and then revert the data when the API call fails. Alternatively, we can remove the `loading: false` from the merge call and add it in a separate merge call after the API call comes back, but that'll be pretty much the same thing we're doing now, so I don't think this is worth updating without a full refactor.

Comment (roryabraham) [reply src/libs/actions/BankAccounts.js:621]: Okay, fair enough. Let's make a note of this in https://github.com/Expensify/Expensify/issues/173363

**Review by roryabraham — APPROVED** [review]

**Review by Julesssss — APPROVED** [review]

**Comment by Julesssss** [comment]
Confirmed the edge case is fixed 👍

**Comment by isagoico** [comment]
@jags96 Hello! We're currently unable to test this add bank account flow with our testing domains since all already have the Expensify card. Can we test this via navigating to staging.new.expensify.com/bank-account/new?

**Comment by isagoico** [comment]
@jags96 Hello! We're currently unable to test this add bank account flow with our testing domains since all already have the Expensify card. Can we test this via navigating to staging.new.expensify.com/bank-account/new?

**Comment by isagoico** [comment]
@jags96 Hello! We're currently unable to test this add bank account flow with our testing domains since all already have the Expensify card. Can we test this via navigating to staging.new.expensify.com/bank-account/new?

**Comment by Jag96** [comment]
@isagoico no worries, I'll do QA for this one and check it off on the spreadsheet!

**Pull request merged**
